code/modules/detective_work/scanner.dm

```diff
@@ -17,77 +17,85 @@
 	actions_types = list(/datum/action/item_action/print_report)
 
 /obj/item/detective_scanner/attack_self(var/mob/user)
-	var/search = input(user, "Enter name, fingerprint or blood DNA.", "Find record", "")
+	var/search = lowertext(input(user, "Enter name, fingerprint or blood DNA.", "Find record", ""))
 
 	if(!search || user.stat || user.incapacitated())
 		return
 
-	search = lowertext(search)
-
 	var/name
 	var/fingerprint = "FINGERPRINT NOT FOUND"
 	var/dna = "BLOOD DNA NOT FOUND"
 
 	// I really, really wish I didn't have to split this into two seperate loops. But the datacore is awful.
 
-	for(var/record in GLOB.data_core.general)
+	for(var/record in GLOB.data_core.general) // Search in the 'general' datacore
 		var/datum/data/record/S = record
-		if(S && (search == lowertext(S.fields["fingerprint"]) || search == lowertext(S.fields["name"])))
+		if(S && (search == lowertext(S.fields["fingerprint"]) || search == lowertext(S.fields["name"]))) // Get Fingerprint and Name
 			name = S.fields["name"]
 			fingerprint = S.fields["fingerprint"]
-			continue
+			break
 
-	for(var/record in GLOB.data_core.medical)
+	for(var/record in GLOB.data_core.medical) // Then search in the 'medical' datacore
 		var/datum/data/record/M = record
-		if(M && ( search == lowertext(M.fields["b_dna"]) || name == M.fields["name"]) )
+		if(M && (search == lowertext(M.fields["b_dna"]) || name == M.fields["name"])) // Get Blood DNA
 			dna = M.fields["b_dna"]
 
-			if(fingerprint == "FINGERPRINT NOT FOUND") // We have searched by DNA, and do not have the relevant information from the fingerprint records.
+			if(fingerprint == "FINGERPRINT NOT FOUND") // We have searched for DNA, and so do not have the relevant information from the fingerprint records.
 				name = M.fields["name"]
 				for(var/gen_record in GLOB.data_core.general)
 					var/datum/data/record/S = gen_record
 					if(S && (name == S.fields["name"]))
 						fingerprint = S.fields["fingerprint"]
-						continue
-			continue
+						break
```

Comment:
I'm really not sure about that whole loop at line 45, but I couldn't figure out how to do it better.
If you search for DNA, then it runs a loop over `data_core.general` twice, which just seems really inefficient.

Of course, every time I've worked on this it's been around midnight, so I might just be missing something obvious.

Reply:
I'd say think about what this two loops try to achieve. Then think okay when is this achieved and is it always required?
From that point you can most likely reconstruct it so that it is more efficient and logical

Reply:
It could be stockholm syndrome talking, but I'm starting to think this works fine as it is.

- If you search for a name or fingerprint, it'll look through the general datacore for them, and then get the blood DNA from the medical datacore, based on the name.

- If you searched for DNA, that's not in the general datacore, so it won't find anything in the first`for()` loop.
Then it gets the DNA and the name, and searches the general datacore again for the fingerprint.

The best way to fix it would be to skip the first loop if you're searching for a fingerprint, but there's no way to tell that automatically. And I don't think it's worth adding a 'Name, Fingerprint or DNA?' alert.